From 2d5a1993e7f9790501c6808b5c4d9e6fe523369f Mon Sep 17 00:00:00 2001
From: Mel Gorman <mgorman@suse.de>
Date: Thu, 6 Feb 2014 10:28:22 +0000
Subject: [PATCH] mm: Warn users of node memory hot-remove if the memory ratio
 is a high risk

References: Memory node hot-remove (fate#316483)
Patch-mainline: Never, upstream developers refusing to acknowledge problem

Warn users of node memory hot-remove is the lowmem:movable ratio is worse
than 1:3.

Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/page_alloc.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 8e2de76..2e44be9 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -3819,6 +3819,31 @@ void __ref build_all_zonelists(pg_data_t *pgdat, struct zone *zone)
 #ifdef CONFIG_NUMA
 	printk("Policy zone: %s\n", zone_names[policy_zone]);
 #endif
+
+	/* Check normal :movable ratio when memory node hot-remove is enabled */
+	if (movable_node_is_enabled()) {
+		unsigned long nr_movable = 0;
+
+		for_each_zone(zone) {
+			if (!populated_zone(zone) || zone_idx(zone) != ZONE_MOVABLE)
+				continue;
+
+			nr_movable += zone->managed_pages;
+		}
+
+		/* Warn if lowmem:movable is worse than 1:3 */
+		if ((vm_total_pages >> 2) < nr_movable) {
+			pr_warn("Memory node hot-remove is enabled but the "
+				"ratio of lowmem:movable memory is very high. "
+				"There may be OOM problems if the kernel "
+				"cannot allocate memory as it is all marked "
+				"for hot-remove. If these problems are "
+				"experienced then boot with a limited amount "
+				"of memory and manually hot-add nodes as "
+				"normal or movable memory\n");
+			WARN_ON_ONCE(1);
+		}
+	}
 }
 
 /*
